[OpenVINO] Enable GPU support for OpenVINO vLLM backend

sshlyapn opened the pull request:
These changes add GPU device support for OpenVINO vLLM backend

- Added `VLLM_OPENVINO_DEVICE` environment variable for OpenVINO device selection
- Updated GPU-related components in OpenVINO backend (KV cache shapes, swap capability, model profiling run etc)
- Updated OpenVINO version to 2024.4 RC1 in dependencies
- Updated installation instructions

Some performance measurements obtained for Intel ARC A770 (16GB) for 1000 prompts from ShareGPT_V3_unfiltered_cleaned_split dataset:

<details>

![image](https://github.com/user-attachments/assets/2eda6376-d472-43a8-8c9a-067d476077a9)

</details>

sshlyapn committed:
[OpenVINO] Add GPU support for OpenVINO backend

sshlyapn committed:
Update OpenVINO version to 2024.5 RC1

sshlyapn committed:
Add GPU device configuration instructions for vLLM OpenVINO backend

liuxingbin commented:
Hi
I tried the PR, but new error occurred. I use openvino-gpu to run qwen2-0.5b. It turns out:
```
Traceback (most recent call last):
  File "/workspace/vllm/vllm/worker/openvino_worker.py", line 302, in determine_num_available_blocks
    kv_cache_size = self.profile_run()
  File "/workspace/vllm/vllm/worker/openvino_worker.py", line 549, in profile_run
    model_profile_run()
  File "/workspace/vllm/vllm/worker/openvino_worker.py", line 538, in model_profile_run
    self.model_runner.execute_model(seqs,
  File "/usr/local/lib/python3.10/dist-packages/torch/utils/_contextlib.py", line 116, in decorate_context
    return func(*args, **kwargs)
  File "/workspace/vllm/vllm/worker/openvino_model_runner.py", line 340, in execute_model
    hidden_states = model_executable(**execute_model_kwargs)
  File "/usr/local/lib/python3.10/dist-packages/nncf/torch/dynamic_graph/wrappers.py", line 146, in wrapped
    return module_call(self, *args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1553, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1562, in _call_impl
    return forward_call(*args, **kwargs)
  File "/workspace/vllm/vllm/model_executor/model_loader/openvino.py", line 164, in forward
    self.ov_request.wait()
RuntimeError: Exception from src/inference/src/cpp/infer_request.cpp:245:
Exception from src/bindings/python/src/pyopenvino/core/infer_request.hpp:54:
Caught exception: Check '!exceed_allocatable_mem_size' failed at src/plugins/intel_gpu/src/runtime/ocl/ocl_engine.cpp:139:
[GPU] Exceeded max size of memory object allocation: requested 19914555392 bytes, but max alloc size supported by device is 1073741824 bytes.Please try to reduce batch size or use lower precision.
```

19914555392 bytes equals 18.5GB. That's strange. I tried some solutions, but they didn't solve my problem.
Any solution or hint?

FYI: I use gpu-version vllm to run qwen2-1.5b, which uses ~8GB by showing nvidia-smi.

liuxingbin commented:
> Hi I tried the PR, but new error occurred. I use openvino-gpu to run qwen2-0.5b. It turns out:
> 
> ```
> Traceback (most recent call last):
>   File "/workspace/vllm/vllm/worker/openvino_worker.py", line 302, in determine_num_available_blocks
>     kv_cache_size = self.profile_run()
>   File "/workspace/vllm/vllm/worker/openvino_worker.py", line 549, in profile_run
>     model_profile_run()
>   File "/workspace/vllm/vllm/worker/openvino_worker.py", line 538, in model_profile_run
>     self.model_runner.execute_model(seqs,
>   File "/usr/local/lib/python3.10/dist-packages/torch/utils/_contextlib.py", line 116, in decorate_context
>     return func(*args, **kwargs)
>   File "/workspace/vllm/vllm/worker/openvino_model_runner.py", line 340, in execute_model
>     hidden_states = model_executable(**execute_model_kwargs)
>   File "/usr/local/lib/python3.10/dist-packages/nncf/torch/dynamic_graph/wrappers.py", line 146, in wrapped
>     return module_call(self, *args, **kwargs)
>   File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1553, in _wrapped_call_impl
>     return self._call_impl(*args, **kwargs)
>   File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1562, in _call_impl
>     return forward_call(*args, **kwargs)
>   File "/workspace/vllm/vllm/model_executor/model_loader/openvino.py", line 164, in forward
>     self.ov_request.wait()
> RuntimeError: Exception from src/inference/src/cpp/infer_request.cpp:245:
> Exception from src/bindings/python/src/pyopenvino/core/infer_request.hpp:54:
> Caught exception: Check '!exceed_allocatable_mem_size' failed at src/plugins/intel_gpu/src/runtime/ocl/ocl_engine.cpp:139:
> [GPU] Exceeded max size of memory object allocation: requested 19914555392 bytes, but max alloc size supported by device is 1073741824 bytes.Please try to reduce batch size or use lower precision.
> ```
> 
> 19914555392 bytes equals 18.5GB. That's strange. I tried some solutions, but they didn't solve my problem. Any solution or hint?
> 
> FYI: I use gpu-version vllm to run qwen2-1.5b, which uses ~8GB by showing nvidia-smi.

change memory [here](https://github.com/sshlyapn/vllm_gpu/blob/ov_gpu_support/vllm/envs.py#L297) to 4gb solver my problem.

sshlyapn commented:
Hi @WoosukKwon, could you please take a look at these changes?

mgoin commented on vllm/executor/openvino_executor.py:
```diff
@@ -143,29 +149,44 @@ def _verify_and_get_model_config(config: ModelConfig) -> ModelConfig:
     kv_cache_space = envs.VLLM_OPENVINO_KVCACHE_SPACE
     if kv_cache_space >= 0:
-        if kv_cache_space == 0:
+        if kv_cache_space == 0 and "CPU" in ov_device:
```
I think it is worth replacing all these `"CPU" in ov_device` checks with a helper function to enforce standards like we currently have with `is_cuda`, `is_hip`, etc

sshlyapn replied:
Yeah, agree, had the same thought but wasn't sure where the best place to define the function would be. I have defined it in `openvino_executor` to make it OpenVINO specific only - what do you think?

mgoin replied:
Yeah I think it is best to put it in openvino files so the executor is a fine place, thanks!

mgoin reviewed (commented):
Overall this seems reasonable, thanks for the updates to logging. Most of my comments are just related to refactoring common checks in helper functions

sshlyapn committed:
Applied review comments:
- Added unified function to check device type
- Added device type check for executor
- Applied other minor fixes

sshlyapn committed:
Bump OpenVINO version to 2024.4.0

sshlyapn commented on vllm/executor/openvino_executor.py:
```diff
@@ -143,29 +149,44 @@ def _verify_and_get_model_config(config: ModelConfig) -> ModelConfig:
+            else:
+                config.cache_dtype = ov.Type.f16
         else:
             config.cache_dtype = ov.Type.f16
```
Currently bf16 data type is not supported on GPU and f16 is used as the primary data type

sshlyapn commented on vllm/executor/openvino_executor.py:
```diff
@@ -143,29 +149,44 @@ def _verify_and_get_model_config(config: ModelConfig) -> ModelConfig:
+    if "CPU" in ov_device:
+        if config.block_size != 32:
+            logger.info(
+                f"OpenVINO optimal block size is 32, overriding currently set {config.block_size}"  # noqa: G004, E501
```
Applied, thanks!

sshlyapn commented on vllm/worker/openvino_worker.py:
```diff
@@ -74,34 +81,100 @@ def __init__(
+                value_blocks = ov.Tensor(self.cache_config.cache_dtype,
+                                         v_block_shape)
+                kv_cache.append((key_blocks, value_blocks))
+        else:
+            # Update key_cache shape:
```
I have added an additional assertion in `_init_executor()` to ensure the device is either CPU or GPU, just to avoid having this assertion in multiple places. Is it okay? It's unlikely that any more device types will be added soon 😊

mgoin replied:
Okay good to know, I agree one assert at the beginning is sufficient

sshlyapn commented:
@mgoin, thank you for your comments! I [applied them](https://github.com/vllm-project/vllm/pull/8192/commits/1723d77e7352d7138b14d1427cc16f1987ef5761) and rebased the branch on top of the recent main, please take a look

mgoin reviewed (approved):
Thanks for the improvements, LGTM just one nit on the documentation

sshlyapn committed:
Add link to the list of supported GPUs

sshlyapn commented on docs/source/getting_started/openvino-installation.rst:
```diff
@@ -3,7 +3,7 @@
 ==========================
 
-vLLM powered by OpenVINO supports all LLM models from :doc:`vLLM supported models list <../models/supported_models>` and can perform optimal model serving on all x86-64 CPUs with, at least, AVX2 support. OpenVINO vLLM backend supports the following advanced vLLM features:
+vLLM powered by OpenVINO supports all LLM models from :doc:`vLLM supported models list <../models/supported_models>` and can perform optimal model serving on all x86-64 CPUs with, at least, AVX2 support, as well as on both integrated and discrete Intel® GPUs (starting from Intel® UHD Graphics generation). OpenVINO vLLM backend supports the following advanced vLLM features:
```
Sure, I have updated the docs with the link to the supported GPU devices: https://docs.openvino.ai/2024/about-openvino/release-notes-openvino/system-requirements.html#gpu

mgoin replied:
Excellent!

mgoin commented:
Thanks for the quick feedback. I am enabling the full CI to run - you may want to merge with latest main to get the CI green since there have been failures before today

p-durandin commented:
@mgoin , @WoosukKwon could you activate automerge?
